pie/utils/objects.py

```diff
@@ -96,3 +96,62 @@ def check(reaction, user):
                     with contextlib.suppress(nextcord.Forbidden):
                         await message.remove_reaction("▶️", user)
                     await message.edit(embed=self.pages[pagenum])
+
+
+class ConfirmView(nextcord.ui.View):
+    """Class for making confirmation embeds easy.
+    The right way of getting response is first calling wait() on instance,
+    then checking instance attribute `value`.
+
+    Attributes:
+        value: True if confirmed, False if declined, None if timed out
+        ctx: Context of command
+        message: Confirmation message
+
+    Args:
+        ctx (:class:`nextcord.ext.commands.Context`): The context for translational and sending purposes.
+        embed (:class:`nextcord.Embed`): Embed to send.
+    """
+
+    def __init__(self, ctx: commands.Context, embed: nextcord.Embed):
+        super().__init__()
+        self.value = None
+        self.ctx = ctx
+        self.embed = embed
+
+    async def send(self):
+        """Sends message to channel defined by command context.
+
+        Returns:
+            True if confirmed, False if declined, None if timed out
+        """
+        self.message = await self.ctx.send(embed=self.embed, view=self)
+        await self.wait()
+        return self.value
+
+    # When the confirm button is pressed, set the inner value to `True` and
+    # stop the View from listening to more input.
+    # We also send the user an ephemeral message that we're confirming their choice.
+    @nextcord.ui.button(label="✅", style=nextcord.ButtonStyle.green)
+    async def confirm(
+        self, button: nextcord.ui.Button, interaction: nextcord.Interaction
+    ):
+        await self._process(True, interaction)
+
+    # This one is similar to the confirmation button except sets the inner value to `False`
+    @nextcord.ui.button(label="❎", style=nextcord.ButtonStyle.red)
+    async def cancel(
+        self, button: nextcord.ui.Button, interaction: nextcord.Interaction
+    ):
+        await self._process(False, interaction)
+
+    async def _process(self, value, interaction):
+        if interaction.user.id is not self.ctx.author.id:
+            return
+
+        try:
+            await self.message.delete()
+        except (nextcord.errors.NotFound, nextcord.errors.Forbidden):
+            pass
+        self.value = value
+        self.stop()
```

Comment:
I came up with this implementation, which would end up being translatable. What do you think?

```py
class ConfirmView(nextcord.ui.View):
    """Class for making confirmation embeds easy.
    The right way of getting response is first calling wait() on instance,
    then checking instance attribute `value`.
    Attributes:
        value: True if confirmed, False if declined, None if timed out
        ctx: Context of command
        message: Confirmation message
    Args:
        ctx: The context for translational and sending purposes.
        embed: Embed to send.
        timeout: Number of seconds before timeout. `None` if no timeout
    """

    def __init__(
        self,
        ctx: commands.Context,
        embed: nextcord.Embed,
        timeout: Union[int, float, None] = 300,
    ):
        super().__init__(timeout=timeout)
        self.value: Optional[bool] = None
        self.ctx = ctx
        self.embed = embed

    async def send(self):
        """Sends message to channel defined by command context.
        Returns:
            True if confirmed, False if declined, None if timed out
        """
        self.add_item(
            nextcord.ui.Button(
                label=_(self.ctx, "Confirm"),
                style=nextcord.ButtonStyle.green,
                custom_id="confirm-button",
            )
        )
        self.add_item(
            nextcord.ui.Button(
                label=_(self.ctx, "Reject"),
                style=nextcord.ButtonStyle.red,
                custom_id="reject-button",
            )
        )
        message = await self.ctx.reply(embed=self.embed, view=self)
        await self.wait()

        if isinstance(self.ctx.message.channel, nextcord.DMChannel):
            self.clear_items()
            await message.edit(embed=self.embed, view=self)
        else:
            try:
                await message.delete()
            except (
                nextcord.errors.HTTPException,
                nextcord.errors.Forbidden,
                nextcord.errors.NotFound,
            ):
                self.clear_items()
                await message.edit(embed=self.embed, view=self)
        return self.value

    async def interaction_check(self, interaction: nextcord.Interaction) -> None:
        """Gets called when interaction with any of the Views buttons happens."""
        if interaction.user.id is not self.ctx.author.id:
            return

        if interaction.data["custom_id"] == "confirm-button":
            self.value = True
        else:
            self.value = False
        self.stop()

    async def on_timeout(self) -> None:
        """Gets called when the view timeouts."""
        self.value = None
        self.stop()
```

Reply:
This is much better, but I think that different behavior in DM's could be confusing. It has to be mentioned in docs or maybe it should be configurable (delete / edit as init parameter)

Reply:
I've implemented your changes and also added that possibility to choose if delete message or not.

Reply:
Sure, we could add an optional bool parameter `strict_cleanup` and split off the cleanup logic to a standalone method that would either delete the message regardless of where it was sent (`strict_cleanup` == True) or keep it in DMChannels since they are private.

I would also recommend to feep that falue to False by default.

Reply:
> I've implemented your changes and also added that possibility to choose if delete message or not.

Well not quite. I did not want the message to keep hanging in a GuildChannel. Only in a DMChannel since they are private by definition. Imagining this as a Reminders feature, I would want to keep the reminders as private as possible, so deleting them when not in a private channel.

Reply:
But that could be handled by Reminder itself as it's decision. This should be, as a core function, highly configurable.

Reply:
We'll see what @sinus-x has to say

Reply:
Personally I'm OK with a command behaving differently in guild and DM contexts.

Reply:
I'm ok with that too, but I think that should be decided by the command, not hard-coded in core.

Reply:
Sinus maked a decision at Discord - let modules decide if delete or not, default behavior is to delete.